In [3]:
%pip install kagglehub
%pip install matplotlib
%pip install ipywidgets

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [4]:
import kagglehub
import email

In [5]:


# Download latest version
path = kagglehub.dataset_download("wcukierski/enron-email-dataset")

print("Path to dataset files:", path)

Path to dataset files: /home/morrissimons/.cache/kagglehub/datasets/wcukierski/enron-email-dataset/versions/2


In [6]:
import pandas as pd

# The Enron dataset typically has a emails.csv file in the dataset
df = pd.read_csv(f"{path}/emails.csv")
df.head()

,file,message
0,allen-p/_sent_mail/1.,Message-ID: <18782981.1075855378110.JavaMail.e...
1,allen-p/_sent_mail/10.,Message-ID: <15464986.1075855378456.JavaMail.e...
2,allen-p/_sent_mail/100.,Message-ID: <24216240.1075855687451.JavaMail.e...
3,allen-p/_sent_mail/1000.,Message-ID: <13505866.1075863688222.JavaMail.e...
4,allen-p/_sent_mail/1001.,Message-ID: <30922949.1075863688243.JavaMail.e...


In [7]:
first_email = df.iloc[1] 
print("File:", first_email['file'])
print("\nMessage:", first_email['message'])


File: allen-p/_sent_mail/10.

Message: Message-ID: <15464986.1075855378456.JavaMail.evans@thyme>
Date: Fri, 4 May 2001 13:51:00 -0700 (PDT)
From: phillip.allen@enron.com
To: john.lavorato@enron.com
Subject: Re:
Mime-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Transfer-Encoding: 7bit
X-From: Phillip K Allen
X-To: John J Lavorato <John J Lavorato/ENRON@enronXgate@ENRON>
X-cc: 
X-bcc: 
X-Folder: \Phillip_Allen_Jan2002_1\Allen, Phillip K.\'Sent Mail
X-Origin: Allen-P
X-FileName: pallen (Non-Privileged).pst

Traveling to have a business meeting takes the fun out of the trip.  Especially if you have to prepare a presentation.  I would suggest holding the business plan meetings here then take a trip without any formal business meetings.  I would even try and get some honest opinions on whether a trip is even desired or necessary.

As far as the business meetings, I think it would be more productive to try and stimulate discussions across the different groups about what is 

In [8]:
print(f"Total number of emails in the dataset: {len(df):,}")

Total number of emails in the dataset: 517,401


In [9]:
# Extract Message-IDs from the message column using string operations
message_ids = df['message'].str.extract(r'Message-ID: <([^>]+)>', expand=False)
print(message_ids.unique()[0])
print(len(message_ids.unique()))

18782981.1075855378110.JavaMail.evans@thyme
517401


In [10]:
print(df['message'][1])

Message-ID: <15464986.1075855378456.JavaMail.evans@thyme>
Date: Fri, 4 May 2001 13:51:00 -0700 (PDT)
From: phillip.allen@enron.com
To: john.lavorato@enron.com
Subject: Re:
Mime-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Transfer-Encoding: 7bit
X-From: Phillip K Allen
X-To: John J Lavorato <John J Lavorato/ENRON@enronXgate@ENRON>
X-cc: 
X-bcc: 
X-Folder: \Phillip_Allen_Jan2002_1\Allen, Phillip K.\'Sent Mail
X-Origin: Allen-P
X-FileName: pallen (Non-Privileged).pst

Traveling to have a business meeting takes the fun out of the trip.  Especially if you have to prepare a presentation.  I would suggest holding the business plan meetings here then take a trip without any formal business meetings.  I would even try and get some honest opinions on whether a trip is even desired or necessary.

As far as the business meetings, I think it would be more productive to try and stimulate discussions across the different groups about what is working and what is not.  Too often the

In [11]:
e = email.message_from_string(df['message'][0])

for items in e.items():
    print(items)


print("="*80)

print("\n" + e.get_payload())

('Message-ID', '<18782981.1075855378110.JavaMail.evans@thyme>')
('Date', 'Mon, 14 May 2001 16:39:00 -0700 (PDT)')
('From', 'phillip.allen@enron.com')
('To', 'tim.belden@enron.com')
('Subject', '')
('Mime-Version', '1.0')
('Content-Type', 'text/plain; charset=us-ascii')
('Content-Transfer-Encoding', '7bit')
('X-From', 'Phillip K Allen')
('X-To', 'Tim Belden <Tim Belden/Enron@EnronXGate>')
('X-cc', '')
('X-bcc', '')
('X-Folder', "\\Phillip_Allen_Jan2002_1\\Allen, Phillip K.\\'Sent Mail")
('X-Origin', 'Allen-P')
('X-FileName', 'pallen (Non-Privileged).pst')

Here is our forecast

 


In [14]:
#view mail by email message id
def view_mail_by_id(message_id):
    mail = df[df['message'].str.contains(message_id)]
    print(mail['message'].values[0])

view_mail_by_id("<15464986.1075855378456.JavaMail.evans@thyme>")

IndexError: index 0 is out of bounds for axis 0 with size 0

In [10]:
def check_payload(message_str):
    msg = email.message_from_string(message_str)
    return bool(msg.get_payload())

def check_payload_length(message_str):
    msg = email.message_from_string(message_str)
    return len(msg.get_payload())

# Apply the function to all messages and count those with payloads
has_payload = df['message'].apply(check_payload)
payload_lengths = df['message'].apply(check_payload_length)
print(f"Emails with payload: {sum(has_payload):,} out of {len(df):,}")
print(f"smallest payload: {min(payload_lengths)}")

Emails with payload: 517,401 out of 517,401
smallest payload: 1


In [11]:
import re

#def find_ssn(message_str):
#    msg = email.message_from_string(message_str)
#    payload = msg.get_payload()
#    pattern = r"^\d{3}-\d{2}-\d{4}$"
#    for line in payload.splitlines():
#        if re.search(pattern, line):
#            return line
#    return None
#
#df['ssn'] = df['message'].apply(find_ssn)


In [12]:
#print(df['ssn'].dropna().unique())
#print(len(df['ssn'].dropna().unique()))

In [13]:
# # Get the unique indices where SSNs were found
# ssn_indices = df['ssn'].dropna().index
# 
# # Display full messages for each email containing an SSN
# for idx in ssn_indices:
#     print(f"\n{'='*80}")
#     print(f"Index: {idx}")
#     print(f"SSN found: {df['ssn'][idx]}")
#     print(f"File: {df['file'][idx]}")
#     print(f"\nFull message:")
#     print(f"{'='*80}")
#     print(df['message'][idx])
#     print(f"{'='*80}\n")

In [14]:
#def find_credit_cards(message_str):
#    msg = email.message_from_string(message_str)
#    payload = msg.get_payload()
#    if not payload:
#        return None
#    
#    patterns = {
#        'Visa': r'4[0-9]{12}(?:[0-9]{3})?',
#        'MasterCard': r'5[1-5][0-9]{14}',
#        'AmEx': r'3[47][0-9]{13}',
#        'Discover': r'6(?:011|5[0-9]{2})[0-9]{12}'
#    }
#    
#    found_cards = {}
#    for card_type, pattern in patterns.items():
#        matches = re.findall(pattern, payload.replace(' ', ''))
#        if matches:
#            found_cards[card_type] = matches
#    
#    return found_cards if found_cards else None
#
## Apply the function to all messages
#df['credit_cards'] = df['message'].apply(find_credit_cards)
#
## Show results
#cards_found = df['credit_cards'].dropna()
#print(f"Found credit card numbers in {len(cards_found)} emails")

In [15]:
# show me the emails with credit card numbers

# for idx, cards in cards_found.items():
#     print(f"\n{'='*80}")
#     print(f"Index: {idx}")
#     print(f"File: {df['file'][idx]}")
#     print(f"\nEmail Payload:")
#     msg = email.message_from_string(df['message'][idx])
#     print(msg.get_payload())
#     print(f"\nCredit card numbers found:")
#     print(f"{'-'*80}")
#     for card_type, card_numbers in cards.items():
#         print(f"{card_type}: {card_numbers}")
#     print(f"{'='*80}")

#### Notes so far
- We got the dataset working 
- 517,401 emails unique emails with their own id.
- Looks like we can get a payload from all the emails with okay content.
- Found a couple of mails with SSN
- Found a couple of mails with creditcard information

##### Conclusion
The mails contain sensetive PII information that we can mask.

## NLP implementation

### presidio test

In [16]:
from presidio_analyzer import AnalyzerEngine

analyzer = AnalyzerEngine()


# Call analyzer to get results
results = analyzer.analyze(text="My phone number is 212-555-5555",
                           entities=["PHONE_NUMBER"],
                           language='en')
print(results)

/home/morrissimons/miniconda3/lib/python3.12/site-packages/transformers/utils/generic.py:441: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  _torch_pytree._register_pytree_node(
/home/morrissimons/miniconda3/lib/python3.12/site-packages/transformers/utils/generic.py:309: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  _torch_pytree._register_pytree_node(


[type: PHONE_NUMBER, start: 19, end: 31, score: 0.75]


In [17]:
def view_email_with_index(df, idx):
    if idx in df.index:
        print(f"Email at index {idx}:")
        print(df.loc[idx, 'message'])
        print("-" * 80)
    else:
        print(f"No email found at index {idx}")

view_email_with_index(df, 495763)


Email at index 495763:
Message-ID: <16610935.1075852880617.JavaMail.evans@thyme>
Date: Sat, 3 Feb 2001 07:15:00 -0800 (PST)
From: infocmb@cuisinemb.com
To: kimberly.watson@enron.com
Subject: Re: Tasting demonstration classes
Mime-Version: 1.0
Content-Type: text/plain; charset=ANSI_X3.4-1968
Content-Transfer-Encoding: 7bit
X-From: La Cuisine de Marie-Blanche  <infocmb@cuisinemb.com>
X-To: Kimberly.Watson <Kimberly.Watson@enron.com>
X-cc: 
X-bcc: 
X-Folder: \KWATSON (Non-Privileged)\Watson, Kimberly\Paris
X-Origin: Watson-K
X-FileName: KWATSON (Non-Privileged).pst

Please respond to infocmbDear Ms. Watson,
We are very much looking forward to meeting you and your group on February 19.

Our address is    18, avenue de La Motte Picquet, 75007 PARIS
Our telepnone     01 45 51 36 34
Our M?tro stop   Ecole militaire
Thank you for the ladies's name and the Visa card or Ms. Janell
Please contact us for more details if needed
Sincerely,
Marie-Blanche de Broglie

Kimberly.Watson@enron.com a *crit 

In [18]:
from presidio_analyzer import AnalyzerEngine
from presidio_anonymizer import AnonymizerEngine
from collections import Counter

def mask_pii_in_email(message_str, entities):
    """
    Takes a raw email message string and a list of PII entity types,
    then analyzes and masks the detected PII (including titles if 'TITLE'
    is in the list of entities).
    """
    # Initialize engines
    analyzer = AnalyzerEngine()
    anonymizer = AnonymizerEngine()
    
    # Get email payload
    msg = email.message_from_string(message_str)
    payload = msg.get_payload()
    
    # Analyze for PII
    analyzer_results = analyzer.analyze(
        text=payload,
        entities=entities,
        language='en'
    )
    
    # Count entities
    entity_counts = Counter(result.entity_type for result in analyzer_results)
    
    # Anonymize the text
    anonymized_text = anonymizer.anonymize(
        text=payload,
        analyzer_results=analyzer_results
    )
    
    return anonymized_text.text, entity_counts

all_base_entities = [
            "PERSON", "PHONE_NUMBER", "EMAIL_ADDRESS", 
            "CREDIT_CARD", "URL", "US_BANK_NUMBER",
            "DATE_TIME", "US_SSN", "US_BANK_NUMBER",
            "US_ITIN", "US_PASSPORT", "LOCATION",
            "IP_ADDRESS", "IBAN_CODE","CRYPTO",
            "MEDICAL_LICENSE"

        ]
# Test it on a specific email
example_masked, counts = mask_pii_in_email(df['message'][495763],all_base_entities)
print("Entity counts:")
for entity, count in counts.items():
    print(f"{entity}: {count}")
print("\nMasked email content:")
print("-" * 80)
print(example_masked)

Entity counts:
EMAIL_ADDRESS: 15
CREDIT_CARD: 1
PERSON: 24
DATE_TIME: 15
LOCATION: 7
URL: 15
US_BANK_NUMBER: 2

Masked email content:
--------------------------------------------------------------------------------
Please respond to infocmbDear Ms. <PERSON>,
We are very much looking forward to meeting you and your group on <DATE_TIME>.

Our address is    18, avenue de La Motte Picquet, <DATE_TIME> <LOCATION>
Our telepnone     01 45 51 <DATE_TIME>
Our M?tro stop   Ecole militaire
Thank you for the ladies's name and the Visa card or Ms. <PERSON>
Please contact us for more details if needed
Sincerely,
<PERSON>

<EMAIL_ADDRESS> a *crit :

> Dear <PERSON>,
>
> We are very excited about our tasting/demonstration class.  Here is a list
> of the ladies attending.  I only have American Express and Master Card, but
> <PERSON> has a VISA and we will put our class fee on her card.
>
> <PERSON>VISA <CREDIT_CARD> exp. <DATE_TIME>  name on card is Janell
> B. Dye
> <PERSON>
> <PERSON>
> <PERSON>
> <P

In [19]:
from tqdm.notebook import tqdm

# all_base_entities = [
#             "PERSON", "PHONE_NUMBER", "EMAIL_ADDRESS", 
#             "CREDIT_CARD", "URL", "US_BANK_NUMBER",
#             "DATE_TIME", "US_SSN", "US_BANK_NUMBER",
#             "US_ITIN", "US_PASSPORT", "LOCATION",
#             "IP_ADDRESS", "IBAN_CODE","CRYPTO",
#             "MEDICAL_LICENSE"
# 
#         ]
# tqdm.pandas()
# 
# # Apply masking to every email message in the dataframe
# def mask_email(message):
#     return mask_pii_in_email(message, all_base_entities)
# 
# # Apply the mask_email function over all messages in the dataframe with a progress bar.
# results = df['message'].progress_apply(mask_email)
# 
# # Create two new columns in df: one for the masked message and one for the entity counts.
# df['masked_message'] = results.apply(lambda res: res[0])
# df['entity_counts'] = results.apply(lambda res: res[1])
# 
# # For inspection, print out the entity counts and masked message for the first 5 emails.
# for idx, row in df[['masked_message', 'entity_counts']].head().iterrows():
#     print(f"Row {idx}:")
#     print("Entity counts:", row['entity_counts'])
#     print("Masked message:\n", row['masked_message'])
#     print("="*80)

In [36]:
# Print all header fields and their values.
print("Headers:")
for header, value in e.items():
    print(f"{header}: {value}")

# Get a specific header, e.g. the 'Subject'
subject = e.get("Subject")
print("\nSubject:", subject)

# Extract the main payload (body) of the email.
payload = e.get_payload()
print("\nPayload:")
print(payload)

print(e.items()[1])
print(e.items()[2])
print(e.items()[3])
print(e.items()[8])
print(e.items()[9])
print(e.items()[11])

print(e.items()[13])

Headers:
Message-ID: <18782981.1075855378110.JavaMail.evans@thyme>
Date: Mon, 14 May 2001 16:39:00 -0700 (PDT)
From: phillip.allen@enron.com
To: tim.belden@enron.com
Subject: 
Mime-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Transfer-Encoding: 7bit
X-From: Phillip K Allen
X-To: Tim Belden <Tim Belden/Enron@EnronXGate>
X-cc: 
X-bcc: 
X-Folder: \Phillip_Allen_Jan2002_1\Allen, Phillip K.\'Sent Mail
X-Origin: Allen-P
X-FileName: pallen (Non-Privileged).pst

Subject: 

Payload:
Here is our forecast

 
('Date', 'Mon, 14 May 2001 16:39:00 -0700 (PDT)')
('From', 'phillip.allen@enron.com')
('To', 'tim.belden@enron.com')
('X-From', 'Phillip K Allen')
('X-To', 'Tim Belden <Tim Belden/Enron@EnronXGate>')
('X-bcc', '')
('X-Origin', 'Allen-P')


### Lessons

#### Mail Index 496,148 from the Mails
- The US driver’s license is not a reliable metric: [Issue #1063](https://github.com/microsoft/presidio/issues/1063). A better alternative is needed.
  - Found instances of `US_DRIVER_LICENSE: 3` where they should have been categorized as credit cards.
- The code still contains minor errors, such as incorrect categorization of places or names:
  - Example: `VIA SANT'ANTONINO 18 - 50123 FIRENZE` (<LOCATION>)
  - Example: `Paul` and `<PERSON>`
  - Spelling error on `09October` for `<DATE_TIME>`.
  - `Our telepnone     01 45 51 36 34` became `Our telepnone     01 45 51 <DATE_TIME>`

#### Mail Index 495,763
- Misclassification of `US_BANK_NUMBER` as a phone number.
- Confusion between `NRP` (Nationality, Religious, or Political group) and `VISA`.
- `US_DRIVER_LICENSE` was mistakenly classified as both a credit card and a phone number.
- Need to mask gendered titles (`Ms.`, `Mr.`) for neutrality.
  - Possible solution: [Deny List Approach](https://microsoft.github.io/presidio/tutorial/01_deny_list/).
  - A better approach might involve replacing titles with gender-neutral alternatives.

TODO: 

Look into what more supported Entities might have problems/are weak. Also look into alternative solutions: https://microsoft.github.io/presidio/supported_entities/

Look into who confidence value can be used for this.

#### Speed Test on My PC
- Processing over 500,000 emails will take approximately **350 hours** on my PC.
- A potential improvement: Implement **batch processing**: [Batch Processing Guide](https://microsoft.github.io/presidio/samples/python/batch_processing/).



In [20]:
## Install and download necessary packages and models if you haven't already:
## !pip install presidio_analyzer presidio_anonymizer
## !python -m spacy download en_core_web_lg
#
## Your PII-masking function for a single email
#def mask_pii_in_email(message_str, entities):
#    """
#    Takes a raw email message string and a list of PII entity types,
#    then analyzes and masks the detected PII.
#    """
#    # Initialize the engines (Note: initializing per email can be slow for large datasets)
#    analyzer = AnalyzerEngine()
#    anonymizer = AnonymizerEngine()
#    
#    # Parse the email to extract its payload (the main text)
#    msg = email.message_from_string(message_str)
#    payload = msg.get_payload()
#    
#    # Analyze the payload for PII
#    analyzer_results = analyzer.analyze(
#        text=payload,
#        entities=entities,
#        language='en'
#    )
#    
#    # Count the detected entities
#    entity_counts = Counter(result.entity_type for result in analyzer_results)
#    
#    # Anonymize the text based on the detected entities
#    anonymized_text = anonymizer.anonymize(
#        text=payload,
#        analyzer_results=analyzer_results
#    )
#    
#    return anonymized_text.text, entity_counts
#
## List of entities you wish to detect and mask.
#all_base_entities = [
#    "PERSON", "PHONE_NUMBER", "EMAIL_ADDRESS", 
#    "CREDIT_CARD", "URL", "US_BANK_NUMBER",
#    "DATE_TIME", "US_SSN", "US_BANK_NUMBER",
#    "US_ITIN", "US_PASSPORT", "LOCATION",
#    "IP_ADDRESS", "IBAN_CODE", "CRYPTO",
#    "MEDICAL_LICENSE"
#]
#
## Enable tqdm progress_apply for DataFrame
#tqdm.pandas()
#
## Apply your masking function on each email message with a progress bar.
#results = df['message'].progress_apply(lambda msg: mask_pii_in_email(msg, all_base_entities))
#
## Create two new columns in the DataFrame:
## - 'masked_message' contains the anonymized email text.
## - 'entity_counts' contains a Counter of the entities detected in that email.
#df['masked_message'] = results.apply(lambda res: res[0])
#df['entity_counts'] = results.apply(lambda res: res[1])
#
#
#for idx, row in df[['masked_message', 'entity_counts']].head().iterrows():
#    print(f"Row {idx}:")
#    print("Entity counts:")
#    for entity, count in row['entity_counts'].items():
#        print(f"  {entity}: {count}")
#    print("=" * 80)
#

  0%|          | 0/517401 [00:00<?, ?it/s]

KeyboardInterrupt: 

In [20]:
# Test title with Deny-list
titles_list = [
    "Sir",
    "Ma'am",
    "Madam",
    "Mr.",
    "Mrs.",
    "Ms.",
    "Miss",
    "Dr.",
    "Professor",
]
gender_list = [
    "he",
    "she",
    "him",
    "her",
    "his",
    "hers",
    "himself",
    "herself",
    "men",
    "gentlemen",
    "guys",
    "boys",
    "bros",
    "male",
    "ladies",
    "lady",
    "women",
    "girls",
    "female",
    "chicks",
]

In [ ]:
from presidio_analyzer import PatternRecognizer

# Create a custom recognizer for TITLE entities using a deny_list.
titles_recognizer = PatternRecognizer(supported_entity="TITLE", deny_list=titles_list)
gender_recognizer = PatternRecognizer(supported_entity="GENDER", deny_list=gender_list)


def mask_pii_in_email_titles_test(message_str, entities):
    """
    Takes a raw email message string and a list of PII entity types,
    then analyzes and masks the detected PII (including titles if 'TITLE'
    is in the list of entities).
    """
    # Initialize engines
    analyzer = AnalyzerEngine()
    anonymizer = AnonymizerEngine()
    
    # Add the custom title recognizer to the analyzer.
    analyzer.registry.add_recognizer(titles_recognizer)
    analyzer.registry.add_recognizer(gender_recognizer)
    
    # Get email payload
    msg = email.message_from_string(message_str)
    payload = msg.get_payload()
    
    # Analyze for PII
    analyzer_results = analyzer.analyze(
        text=payload,
        entities=entities,
        language='en'
    )
    
    # Count entities
    entity_counts = Counter(result.entity_type for result in analyzer_results)
    
    # Anonymize the text
    anonymized_text = anonymizer.anonymize(
        text=payload,
        analyzer_results=analyzer_results
    )
    
    return anonymized_text.text, entity_counts


all_base_entities = [
    "PERSON", "PHONE_NUMBER", "EMAIL_ADDRESS", 
    "CREDIT_CARD", "URL", "US_BANK_NUMBER",
    "DATE_TIME", "US_SSN", "US_ITIN",
    "US_PASSPORT", "LOCATION", "IP_ADDRESS",
    "IBAN_CODE", "CRYPTO", "MEDICAL_LICENSE",
    "TITLE", "GENDER"
]


masked_email, counts = mask_pii_in_email_titles_test(df['message'][495763],all_base_entities)

print("Entity counts:")
for entity, count in counts.items():
    print(f"{entity}: {count}")

print("\nMasked email content:")
print("-" * 80)
print(masked_email)

### Notes and Lessons
Notes

- Genders and Titles are now part of the masking process.

Lessons

- Batching gave us a speedup of approx 2,6 X and a run time of 134 Hours.
  - Still to big, we might sample out 2,500 random mails to preform some tests. 

TODO:
- Test Lemmatization and Stemming. Hypothesis Lemmatization is best in Microsoft Presidio and will give a performance boost.
  - to test the performance boost we need a ground truth dataset. for this we will use SOTA PII-Masking tools



